Build failure: openocd

### Steps To Reproduce

Steps to reproduce the behavior:
1. `nix build .#openocd`

### Build log

```
...
src/jtag/drivers/linuxgpiod.c: In function 'linuxgpiod_read':
src/jtag/drivers/linuxgpiod.c:51:18: warning: implicit declaration of function 'gpiod_line_get_value'; did you mean 'gpiod_line_request_get_value'? [-Wimplici>
   51 |         retval = gpiod_line_get_value(gpiod_line[ADAPTER_GPIO_IDX_TDO]);
      |                  ^~~~~~~~~~~~~~~~~~~~
      |                  gpiod_line_request_get_value
src/jtag/drivers/linuxgpiod.c: In function 'linuxgpiod_write':
src/jtag/drivers/linuxgpiod.c:84:26: warning: implicit declaration of function 'gpiod_line_set_value'; did you mean 'gpiod_line_request_set_value'? [-Wimplici>
   84 |                 retval = gpiod_line_set_value(gpiod_line[ADAPTER_GPIO_IDX_TDI], tdi);
      |                          ^~~~~~~~~~~~~~~~~~~~
      |                          gpiod_line_request_set_value
...skipping...
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~
src/jtag/drivers/linuxgpiod.c:351:18: error: 'struct gpiod_line_request_config' has no member named 'consumer'
  351 |                 .consumer = "OpenOCD",
      |                  ^~~~~~~~
src/jtag/drivers/linuxgpiod.c:351:29: warning: excess elements in struct initializer
  351 |                 .consumer = "OpenOCD",
      |                             ^~~~~~~~~
src/jtag/drivers/linuxgpiod.c:351:29: note: (near initialization for 'config')
src/jtag/drivers/linuxgpiod.c:352:18: error: 'struct gpiod_line_request_config' has no member named 'request_type'
  352 |                 .request_type = dir,
      |                  ^~~~~~~~~~~~
src/jtag/drivers/linuxgpiod.c:352:33: warning: excess elements in struct initializer
  352 |                 .request_type = dir,
      |                                 ^~~
src/jtag/drivers/linuxgpiod.c:352:33: note: (near initialization for 'config')
src/jtag/drivers/linuxgpiod.c:353:18: error: 'struct gpiod_line_request_config' has no member named 'flags'
  353 |                 .flags = flags,
      |                  ^~~~~
src/jtag/drivers/linuxgpiod.c:353:26: warning: excess elements in struct initializer
  353 |                 .flags = flags,
      |                          ^~~~~
src/jtag/drivers/linuxgpiod.c:353:26: note: (near initialization for 'config')
src/jtag/drivers/linuxgpiod.c:350:42: error: storage size of 'config' isn't known
  350 |         struct gpiod_line_request_config config = {
      |                                          ^~~~~~
src/jtag/drivers/linuxgpiod.c:356:18: warning: implicit declaration of function 'gpiod_line_request'; did you mean 'gpiod_line_request_get_fd'? [-Wimplicit-fu>
  356 |         retval = gpiod_line_request(gpiod_line[idx], &config, val);
      |                  ^~~~~~~~~~~~~~~~~~
      |                  gpiod_line_request_get_fd
src/jtag/drivers/linuxgpiod.c:350:42: warning: unused variable 'config' [-Wunused-variable]
  350 |         struct gpiod_line_request_config config = {
      |                                          ^~~~~~
make[2]: *** [Makefile:3707: src/jtag/drivers/libocdjtagdrivers_la-linuxgpiod.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I./src -I./src -DPKGDATADIR=\"/nix/store/8l91pvwv2rr2d11ib8xpx15kb2waj09b-openocd-0.12.0/share/openocd\" -DBINDIR=>
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I./src -I./src -DPKGDATADIR=\"/nix/store/8l91pvwv2rr2d11ib8xpx15kb2waj09b-openocd-0.12.0/share/openocd\" -DBINDIR=>
make[2]: Leaving directory '/build/openocd-0.12.0'
make[1]: *** [Makefile:5013: all-recursive] Error 1
make[1]: Leaving directory '/build/openocd-0.12.0'
make: *** [Makefile:2113: all] Error 2
```

### Additional context

Add any other context about the problem here.

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->
CC @bjornfor @prusnak 
### Metadata


```console
> nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 6.4.4, NixOS, 23.11 (Tapir), 23.11.20230729.e6ab469`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.15.1`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

--- a/pkgs/development/embedded/openocd/default.nix
+++ b/pkgs/development/embedded/openocd/default.nix
@@ -25,7 +25,15 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [ pkg-config ];
 
   buildInputs = [ hidapi jimtcl libftdi1 libjaylink libusb1 ]
-    ++ lib.optional stdenv.isLinux libgpiod;
+    ++
+    # tracking issue for v2 api changes https://sourceforge.net/p/openocd/tickets/306/
+    lib.optional stdenv.isLinux (libgpiod.overrideAttrs (old: rec {
+      version = "1.6.4";
+      src = fetchurl {
+        url = "https://git.kernel.org/pub/scm/libs/libgpiod/libgpiod.git/snapshot/libgpiod-${version}.tar.gz";
+        sha256 = "sha256-gp1KwmjfB4U2CdZ8/H9HbpqnNssqaKYwvpno+tGXvgo=";
+      };
+    }));
 
   configureFlags = [
     "--disable-werror"

